[JIT] SubgraphRewriter: Expose runOnGraph and use it in tests.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #20084 [JIT] SubgraphRewriter: Add a support for arbitrary replacement graphs in subgraph rewriter.
* **#20083 [JIT] SubgraphRewriter: Expose runOnGraph and use it in tests.**
* #20082 [JIT] SubgraphRewriter: Rename pattern fusion to subgraph rewrite.

Differential Revision: [D15190192](https://our.internmc.facebook.com/intern/diff/D15190192)

Commit:
[JIT] SubgraphRewriter: Expose runOnGraph and use it in tests.

raghuramank100 commented on test/test_jit.py:
```diff
@@ -1440,56 +1440,63 @@ def forward(self, x):
-                return o
-        m = testModule()
-        torch._C._jit_pass_custom_pattern_based_rewrite("""
+        # mul(mul(mul(mul(x,y),z),x),y) --> mul(mul(mulmul(x,y,z), x), y) -->
```
From a user point of view, it is useful to be able to define a pattern in python for matching. Can we have a test where both the source graph and the pattern match are both in python and are then scripted to be converted into IR for matching?

ZolotukhinM replied:
This pass is not for users, it is for us, compiler-writers. Users should not be using it, and if some power users still decide to do that, they should be prepared to work with IR.

Having a python example is deceiving: while it might look nice and easy, one would actually not know the exact pattern they get from it. When it doesn't work (e.g. I bet you would not be able to easily match conv-relu if you write the pattern in python), you would have to debug it and the first step would be to see the pattern in IR.

raghuramank100 reviewed (changes requested):
Please see comment, thanks

Commit:
Update on "[JIT] SubgraphRewriter: Expose runOnGraph and use it in tests."

[JIT] SubgraphRewriter: Expose runOnGraph and use it in tests.

gh-metadata: pytorch pytorch 20083 gh/ZolotukhinM/81/head

Commit:
Update on "[JIT] SubgraphRewriter: Expose runOnGraph and use it in tests."

[JIT] SubgraphRewriter: Expose runOnGraph and use it in tests.

gh-metadata: pytorch pytorch 20083 gh/ZolotukhinM/81/head
